test/functional/lib/ipfs-request-gateway-redirect.test.js

```diff
@@ -216,49 +309,67 @@ describe('modifyRequest.onBeforeRequest:', function () {
 
-        // X-Ipfs-Path to ensure value from URL takes a priority
-        request.responseHeaders = [{ name: 'X-Ipfs-Path', value: fakeXIpfsPathHdrVal }]
-
-        /// We expect redirect to path-based gateway because go-ipfs >=0.5 will
+        // We expect redirect to path-based gateway because go-ipfs >=0.5 will
         // return redirect to a subdomain, and we don't want to break users
         // running older versions of go-ipfs by loading subdomain first and
         // failing.
-        expect((await modifyRequest.onBeforeRequest(request)).redirectUrl)
-          .to.equal(`http://localhost:8080/ipfs/${cid}/`)
+        ensureCallRedirected({
+          modifiedRequestCallResp: await modifyRequest.onBeforeRequest(request),
+          MV2Expectation: `http://localhost:8080/ipfs/${cid}/`,
+          MV3Expectation: {
+            origin: `^https?\\:\\/\\/${cid}\\.ipfs\\.dweb\\.link`,
+            destination: `http://localhost:8080/ipfs/${cid}`
+          }
+        })
       })
       it('should be redirected to localhost gateway (*.ipfs on 3rd party gw)', async function () {
         state.redirect = true
         const request = url2request(`https://${cid}.ipfs.cf-ipfs.com/`)
-        request.responseHeaders = [{ name: 'X-Ipfs-Path', value: fakeXIpfsPathHdrVal }]
-        expect((await modifyRequest.onBeforeRequest(request)).redirectUrl)
-          .to.equal(`http://localhost:8080/ipfs/${cid}/`)
+
+        ensureCallRedirected({
+          modifiedRequestCallResp: await modifyRequest.onBeforeRequest(request),
+          MV2Expectation: `http://localhost:8080/ipfs/${cid}/`,
+          MV3Expectation: {
+            origin: `^https?\\:\\/\\/${cid}\\.ipfs\\.cf\\-ipfs\\.com`,
+            destination: `http://localhost:8080/ipfs/${cid}`
+          }
+        })
       })
       it('should be redirected to localhost gateway and keep URL encoding of original path', async function () {
         state.redirect = true
         const request = url2request('https://bafybeigfejjsuq5im5c3w3t3krsiytszhfdc4v5myltcg4myv2n2w6jumy.ipfs.dweb.link/%3Ffilename=test.jpg?arg=val')
-        request.responseHeaders = [{ name: 'X-Ipfs-Path', value: fakeXIpfsPathHdrVal }]
-        expect((await modifyRequest.onBeforeRequest(request)).redirectUrl)
-          .to.equal('http://localhost:8080/ipfs/bafybeigfejjsuq5im5c3w3t3krsiytszhfdc4v5myltcg4myv2n2w6jumy/%3Ffilename=test.jpg?arg=val')
+
+        ensureCallRedirected({
+          modifiedRequestCallResp: await modifyRequest.onBeforeRequest(request),
+          MV2Expectation: 'http://localhost:8080/ipfs/bafybeigfejjsuq5im5c3w3t3krsiytszhfdc4v5myltcg4myv2n2w6jumy/%3Ffilename=test.jpg?arg=val',
+          MV3Expectation: {
+            origin: '^https?\\:\\/\\/bafybeigfejjsuq5im5c3w3t3krsiytszhfdc4v5myltcg4myv2n2w6jumy\\.ipfs\\.dweb\\.link',
+            destination: 'http://localhost:8080/ipfs/bafybeigfejjsuq5im5c3w3t3krsiytszhfdc4v5myltcg4myv2n2w6jumy'
```

Comment:
:warning:  Not blocking this PR, but see https://github.com/ipfs/ipfs-companion/issues/1255 for follow-up work.